File: c++_programming/all_programs/reverse-linkedlist.cpp
#include <iostream>
using namespace std;
struct node
{
	int data;
	struct node *next;
};

struct node *head=NULL;

void insert_node(int data)
{
	struct node *newnode=(struct node*)malloc(sizeof(struct node));
	newnode->data=data;
	newnode->next=NULL;
	if(head==NULL)
	{
		head=newnode;
		return;
	}
	struct node *current=head;
	while(current->next!=NULL)
		current=current->next;
	current->next=newnode;
	return;
}

void reverse_list()
{

	struct node *prev=NULL,*curr=head,*temp=NULL;
	while(curr)
	{
		temp=curr->next;
		curr->next=prev;
		prev=curr;
		curr=temp;
	}
	head=prev;
	
}
void display()
{
	struct node *current=head;
	while(current)
	{
		cout<<current->data<<endl;
		current=current->next;
	}
}

int main()
{
	insert_node(10);
	insert_node(20);
	insert_node(30);
	insert_node(100);
	display();
	cout<<"After reversing the list"<<endl;

	reverse_list();
	display();
}

File: python/reverse_linkedlist.py
class Node:
    def __init__(self,data,next=None):
        self.data=data
        self.next=next

class Linkedlist:
    def __init__(self):
        self.head=None

def add_node(head,data):
    newnode=Node(data)
    if head is None:
        head=newnode
    else:

        n=head
        while n.next is not None:
            n=n.next

        n.next=newnode
    return head

def print_node(l):
    curr=l
    while(curr):
        print(curr.data)
        curr=curr.next

def reverse_list(l):
    prev=None
    curr=l
    while curr:
        temp=curr.next
        curr.next=prev
        prev=curr
        curr=temp

    head=prev
    return head
    


mylist=Linkedlist()
mylist.head=add_node(mylist.head,10)
add_node(mylist.head,20)
add_node(mylist.head,30)
mylist.head=reverse_list(mylist.head)
print_node(mylist.head)



